File: Dong/Dong/Sections/SLiderMenuViewController.m
//
//  FirstViewController.m
//  Dong
//
//  Created by 姚智豪 on 2018/6/13.
//  Copyright © 2018年 姚智豪. All rights reserved.
//
#import "SLiderMenuViewController.h"
#import <Masonry.h>
#import "UtilsMacro.h"

@interface SLiderMenuViewController ()

@property (weak, nonatomic) IBOutlet UIView *user;
@property (weak, nonatomic) IBOutlet UIView *Home;

@property(nonatomic,strong)IBOutlet UIButton *maskButton;
@property(nonatomic,assign) BOOL folder;
@property(nonatomic,assign) CGPoint originalPoint;
@property(nonatomic,assign) CGFloat scaleFactor;
@property(nonatomic,assign) CGFloat offsetX;
@property(nonatomic,assign) CGFloat originalX;
@property(nonatomic,assign) CGFloat deltaScaleFactor;
@property(nonatomic,assign) CGFloat deltaAlphaFactor;
@property(nonatomic,assign) CGFloat miniumOffestX;

@end

@implementation SLiderMenuViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self initParams];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

-(void)switchController{
    _folder = !_folder;
    if (_folder) {
        [self showHomeViewController];
    }else{
        [self showUserViewController];
    }
}

- (IBAction)panned:(UIPanGestureRecognizer *)sender {
    //locationInView:返回计算为手势位置的点
    CGPoint point = [sender locationInView:self.view];
    switch (sender.state) {
        case UIGestureRecognizerStateBegan:
            _originalPoint = point;
            [self setShadowVisible:YES];
            break;
        case UIGestureRecognizerStateChanged:{
            CGFloat tx = point.x-self.originalPoint.x+self.originalX;
            tx = tx < 0 ? 0 :tx;

            CGFloat s = 1-tx*self.deltaScaleFactor;
            s = s < 0 ? 0 : s;

            CGAffineTransform transform = CGAffineTransformMakeTranslation(tx,0);
            transform = CGAffineTransformScale(transform, s, s);
            _Home.transform = transform;

            CGFloat alaph = tx*self.deltaAlphaFactor;
            alaph=alaph > 1 ? 1 : alaph;
            _user.alpha=alaph;
            }break;
        case UIGestureRecognizerStateEnded:
        case UIGestureRecognizerStateCancelled:{
            CGAffineTransform transform = _Home.transform;
            if (transform.tx >= _miniumOffestX) {
                self.folder = NO ;
                [self showUserViewController];
            }else{
                self.folder = YES ;
                [self showHomeViewController];
            }
        }
        default:
            NSLog(@"other");
            break;
    }
}

-(CGFloat)originalX{
    return (_folder? 0 : _offsetX);
}

#pragma MARK - Private

-(void)initParams{
    //默认折叠
    _folder = YES;
    //缩小
    _scaleFactor=0.95;
    //最终偏移量
    _offsetX = self.view.frame.size.width/2+30;
    //滑动透明度的变化
    _deltaAlphaFactor=1/_offsetX;
    //最小滑动偏移量
    _miniumOffestX=_offsetX/2;
    _deltaScaleFactor=(1-_scaleFactor)/_offsetX;
}

-(void)showUserViewController{
    [self setShadowVisible:YES];
    CGAffineTransform transform = CGAffineTransformMakeTranslation(_offsetX, 0);
    transform = CGAffineTransformScale(transform, _scaleFactor, _scaleFactor);
    [UIView animateWithDuration:kAnimationDuration animations:^{
        self.user.alpha=1;
        self.Home.transform=transform;
    } completion:^(BOOL finished){
        [self setMaskButtonVisible:YES];
    }];
}

-(void)showHomeViewController{
    [self setShadowVisible:NO];
    [UIView animateWithDuration:kAnimationDuration animations:^{
        self.user.alpha=0;
        self.Home.transform=CGAffineTransformIdentity;
    }completion:^(BOOL finished){
        [self setMaskButtonVisible:NO];
    }];
}

//设置阴影
-(void)setShadowVisible:(BOOL)visible{
    if (visible) {
        UIBezierPath *path = [UIBezierPath bezierPathWithRect:_Home.layer.bounds];
        _Home.layer.shadowPath=path.CGPath;
        _Home.layer.shadowColor = [UIColor blackColor].CGColor;
        _Home.layer.shadowOffset=CGSizeZero;
        _Home.layer.shadowOpacity=0.4;
        _Home.layer.shadowRadius=4.5;
    }else{
        _Home.layer.shadowRadius=0;
    }
}

//点击home界面弹回
-(void)setMaskButtonVisible:(BOOL)visible{
    if (_maskButton==nil) {
        self.maskButton = [UIButton buttonWithType:UIButtonTypeCustom];
        [_maskButton addTarget:self action:@selector(maskButtonPressed) forControlEvents:UIControlEventTouchUpInside];
        [_Home addSubview:_maskButton];
        [_maskButton mas_makeConstraints:^(MASConstraintMaker *make){
            make.edges.equalTo(self.Home).insets(UIEdgeInsetsZero);
        }];
    }
    _maskButton.hidden = !visible;
}

-(void)maskButtonPressed{
    _folder=!_folder;
    [self showHomeViewController];
}


@end

File: Dongqiu/Dongqiu/首页/中间界面/SLiderMenuViewController.swift
//
//  SLiderMenuViewController.swift
//  Dongqiu
//
//  Created by 姚智豪 on 2019/1/29.
//  Copyright © 2019年 YZH. All rights reserved.
//

import UIKit

class SLiderMenuViewController: UIViewController {

    var mainContainer: UIView?
    var contantContainer:UIView?
    
    var folder:Bool?
    var originalPoint:CGPoint?
    var scaleFactor:CGFloat?
    var offsetX:CGFloat?
    
    func originalX() -> CGFloat {
        return folder! ? 0 : offsetX!
    }

    var deltaScaleFactor:CGFloat?
    var deltaAlphaFactor:CGFloat?
    var miniumOffestX:CGFloat?
    
    lazy var pan : UIPanGestureRecognizer = {
        let pan = UIPanGestureRecognizer()
        return pan;
    }()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.view.backgroundColor = UIColor.clear
        initParams()
        mainContainer = addChildViewController(viewController: "ContentViewController", isNavViewController: true)
//        contantContainer = addChildViewController(viewController: "ContentViewController", isNavViewController: false)
//        contantContainer?.frame = CGRect(x: -self.view.bounds.size.width*4/5, y: 0, width: self.view.bounds.size.width*4/5, height: self.view.bounds.size.height)
//        let pan = UIPanGestureRecognizer()
//        pan.addTarget(self, action: #selector(panned(sender:)))
//        self.view.addGestureRecognizer(pan)
    }
    
    func addChildViewController(viewController: String,isNavViewController:Bool) -> UIView{
        let nameSpace = Bundle.main.infoDictionary?["CFBundleExecutable"] as! String
        let cls : AnyClass = NSClassFromString(nameSpace + "." + viewController)!
        let vcCls = cls as! UIViewController.Type
        let vc = vcCls.init()
        if isNavViewController {
            let nav = NavViewController(rootViewController: vc)
            self.addChildViewController(nav)
            self.view.addSubview(nav.view)
            return nav.view
        }else{
            self.addChildViewController(vc)
            self.view.addSubview(vc.view)
            return vc.view
        }
    }
    
    func initParams(){
        //默认折叠
        folder = true;
        //
        scaleFactor=0.95;
        //最终偏移量
        offsetX = self.view.frame.size.width*4/5;
        //滑动透明度的变化
        deltaAlphaFactor=1/offsetX!;
        //最小滑动偏移量
        miniumOffestX=offsetX!/2;
        deltaScaleFactor=(1-scaleFactor!)/offsetX!;
    }

    func switchController(){
        folder = !folder!
        if folder!{
            self.showContantViewController()
        }
        else{
            self.showMainViewController()
        }
    }
    
    func showContantViewController(){
        let transform = CGAffineTransform.init(translationX: offsetX!, y: 0)
        UIView.animate(withDuration: 0.35) {
            self.mainContainer?.alpha = 0.5
            self.contantContainer?.transform = transform
        }
    }
    
    func showMainViewController(){
        UIView.animate(withDuration: 0.35) {
            self.mainContainer?.alpha = 1
            self.contantContainer?.transform = CGAffineTransform.identity
        }
    }

    @objc func panned(sender:UIPanGestureRecognizer){
        let point = sender.location(in: self.view)
        switch sender.state {
        case UIGestureRecognizerState.began:
            originalPoint = point
            break
        case .changed:
            var tx = point.x - originalPoint!.x + self.originalX()
            tx = tx < 0 ? 0 : tx
            
            tx = Float(tx) < Float(offsetX!) ? tx : offsetX!
            let transform = CGAffineTransform.init(translationX: tx, y: 0)
            contantContainer?.transform = transform
            break
        case .ended:
            let transform = contantContainer?.transform
            if Float((transform?.tx)!) >= Float(self.miniumOffestX!) {
                self.showContantViewController()
                
            }else{
                self.showMainViewController()
            }
            break
          
        default:
            print("other")
        }
    }
    
   
    
}
